File: app/src/main/java/at/shockbytes/corey/ui/fragment/dialogs/WorkoutMessageDialogFragment.java
package at.shockbytes.corey.ui.fragment.dialogs;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;

import at.shockbytes.corey.R;

/**
 * @author Martin Macheiner
 *         Date: 26.02.2017.
 */

public class WorkoutMessageDialogFragment extends DialogFragment {

    public enum MessageType {QUIT, DONE}

    public interface OnMessageAgreeClickedListener {

        void onMessageAgreeClicked();
    }

    public static final String ARG_MSG_TYPE = "arg_message_type";

    public static WorkoutMessageDialogFragment newInstance(MessageType type) {
        WorkoutMessageDialogFragment fragment = new WorkoutMessageDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_MSG_TYPE, type);
        fragment.setArguments(args);
        return fragment;
    }

    private MessageType msgType;

    private OnMessageAgreeClickedListener listener;

    public WorkoutMessageDialogFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        msgType = (MessageType) getArguments().getSerializable(ARG_MSG_TYPE);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        return new AlertDialog.Builder(getContext())
                .setIcon(R.mipmap.ic_launcher)
                .setTitle(getTitle())
                .setMessage(getMessage())
                .setCancelable(false)
                .setPositiveButton(getPositiveText(), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {

                        if (listener != null) {
                            listener.onMessageAgreeClicked();
                        }
                        dismiss();
                    }
                })
                .setNegativeButton(getNegativeText(), null)
                .create();
    }

    public void setOnMessageAgreeClickedListener(OnMessageAgreeClickedListener listener) {
        this.listener = listener;
    }

    private int getTitle() {
        return (msgType == MessageType.QUIT)
                ? R.string.workout_message_title_quit
                : R.string.workout_message_title_done;
    }

    private int getMessage() {
        return (msgType == MessageType.QUIT)
                ? R.string.workout_message_msg_quit
                : R.string.workout_message_msg_done;
    }

    private int getPositiveText() {
        return (msgType == MessageType.QUIT)
                ? R.string.workout_message_pos_quit
                : R.string.workout_message_pos_done;
    }

    private int getNegativeText() {
        return (msgType == MessageType.QUIT)
                ? R.string.workout_message_neg_quit
                : R.string.workout_message_neg_done;
    }

}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/dialog/WorkoutMessageDialogFragment.kt
package at.shockbytes.corey.ui.fragment.dialog

import android.app.Dialog
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.appcompat.app.AlertDialog
import at.shockbytes.corey.R

/**
 * @author Martin Macheiner
 * Date: 26.02.2017.
 */

class WorkoutMessageDialogFragment : DialogFragment() {

    enum class MessageType {
        QUIT, DONE
    }

    private var msgType: MessageType? = null

    private var listener: (() -> Unit)? = null

    private val title: Int
        get() = if (msgType == MessageType.QUIT)
            R.string.workout_message_title_quit
        else
            R.string.workout_message_title_done

    private val message: Int
        get() = if (msgType == MessageType.QUIT)
            R.string.workout_message_msg_quit
        else
            R.string.workout_message_msg_done

    private val positiveText: Int
        get() = if (msgType == MessageType.QUIT)
            R.string.workout_message_pos_quit
        else
            R.string.workout_message_pos_done

    private val negativeText: Int
        get() = if (msgType == MessageType.QUIT)
            R.string.workout_message_neg_quit
        else
            R.string.workout_message_neg_done

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        msgType = arguments!!.getSerializable(ARG_MSG_TYPE) as MessageType
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        return AlertDialog.Builder(context!!)
                .setIcon(R.mipmap.ic_launcher)
                .setTitle(title)
                .setMessage(message)
                .setCancelable(false)
                .setPositiveButton(positiveText) { _, _ ->
                    listener?.invoke()
                    dismiss()
                }
                .setNegativeButton(negativeText, null)
                .create()
    }

    fun setOnMessageAgreeClickedListener(listener: () -> Unit): WorkoutMessageDialogFragment {
        this.listener = listener
        return this
    }

    companion object {

        private const val ARG_MSG_TYPE = "arg_message_type"

        fun newInstance(type: MessageType): WorkoutMessageDialogFragment {
            val fragment = WorkoutMessageDialogFragment()
            val args = Bundle()
            args.putSerializable(ARG_MSG_TYPE, type)
            fragment.arguments = args
            return fragment
        }
    }
}
